# The Single-Cell Spatial Transcriptomics Analysis (ScSTA) Cookbook
## Organizer: Arun Das
## Affiliation: University of Pittsburgh
## Link: [BHI 2023 Spatial Transcriptomics Tutorial](https://bhi.embs.org/2023/program/tutorials/)

## Abstract:
This tutorial aims to provide attendees with comprehensive knowledge and hands-on experience in the data analysis of sub-cellular resolution single-cell spatial transcriptomics (ScST) technologies such as Nanostring CosMx and 10X Genomics Xenium. ScST is a cutting-edge technology that profiles single-cell transcriptomics with spatial information of cells/transcripts in intact tissues. While it enables researchers to understand gene expression patterns in the context of tissue structure and disease pathology, it poses unique challenges in data analysis. This tutorial will cover each step of the analysis workflow and showcase typical use cases. It intends to accelerate the adoption of spatial single-cell transcriptomics analysis and foster breakthroughs in our understanding of complex biological systems.

## Objectives:
Introduce participants to the principles and significance of ScST.
Familiarize attendees with the components of the Single-Cell Spatial Transcriptomics Analysis (ScSTA) pipeline, including image registration, cell segmentation, cell type identification, spatial gene expression analysis, spatial correlation analysis, and more.
Provide practical experience in using relevant computational tools and software for ScST.
Illustrate the application of ScST in disease pathology research, biomarker discovery, and patient outcome prediction.
Proposed Agenda:

### Module 1: Introduction to Single-Cell Spatial Transcriptomics technology
Overview of single-cell spatial transcriptomics and its limitations.
Importance of incorporating spatial information for deeper insights.
Introduction to the ScSTA pipeline and its key components.

### Module 2: Image Registration and Cell Segmentation
Familiarizing the popular libraries for image registration and cell segmentation.
Importance of maintaining good image registration and cell segmentation in whole-slide-image (WSI) tissues.

### Module 3: Data Preprocessing, Quality Control, and Cell Typing
Cleaning and normalization of the segmented single-cell spatial data.
Addressing technical variability and batch effects.
Quality control measures for reliable downstream analysis.
Using a scRNA-seq reference genomic profiles to carry out cell type transfer to spatial single-cell transcriptomics datasets.
Using marker-based cell typing for ScST datasets.

### Module 4: Spatial Mapping, Neighborhood Analysis, and Visualization
Extracting spatial features such as cell density, average gene expression, and neighborhood cell type composition.
Identifying spatial hotspots using spatial correlation methods.
Developing visualization techniques for spatial mapping of the extracted spatial features.

### Module 5: Hands-on Practical
Guided tutorial using real-world single cell spatial transcriptomics dataset.
Step-by-step walkthrough of the ScST analysis pipeline.
Interactive assistance and troubleshooting.

### Module 6: Future Directions and Challenges
Emerging trends in single-cell spatial omics technologies.
Ongoing challenges and opportunities in the field.
Collaboration possibilities and resources for continued learning.
